provider/pkg/autoscaling/autoscaling.go

```diff
@@ -0,0 +1,88 @@
+// Copyright 2016-2024, Pulumi Corporation.
+//
+// Licensed under the Apache License, Version 2.0 (the "License");
+// you may not use this file except in compliance with the License.
+// You may obtain a copy of the License at
+//
+//     http://www.apache.org/licenses/LICENSE-2.0
+//
+// Unless required by applicable law or agreed to in writing, software
+// distributed under the License is distributed on an "AS IS" BASIS,
+// WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+// See the License for the specific language governing permissions and
+// limitations under the License.
+
+package autoscaling
+
+import (
+	"github.com/hashicorp/terraform-plugin-sdk/v2/helper/schema"
+	"github.com/pulumi/pulumi/sdk/v3/go/common/util/contract"
+)
+
+// Customize resources in the autoscaling module.
+func ReconfigureResources(p *schema.Provider) {
+	autoscalingGroupReconfigure(mustGetResource(p, "aws_autoscaling_group"))
+}
+
+func autoscalingGroupReconfigure(r *schema.Resource) {
+	suppressUndefinedToDefaultDiffs(mustGetField(r, "force_delete"))
+	suppressUndefinedToDefaultDiffs(mustGetField(r, "force_delete_warm_pool"))
+	suppressUndefinedToDefaultDiffs(mustGetField(r, "ignore_failed_scaling_activities"))
+	suppressUndefinedToDefaultDiffs(mustGetField(r, "wait_for_capacity_timeout"))
+}
+
+func mustGetResource(p *schema.Provider, resourceName string) *schema.Resource {
+	res, ok := p.ResourcesMap[resourceName]
+	contract.Assertf(ok, "expected to find the %q resource", resourceName)
+	return res
+}
+
+func mustGetField(r *schema.Resource, fieldName string) *schema.Schema {
+	fd, ok := r.Schema[fieldName]
+	contract.Assertf(ok, "expected to find the %q field", fieldName)
+	return fd
+}
+
+func suppressUndefinedToDefaultDiffs(field *schema.Schema) {
+	contract.Assertf(field.DiffSuppressFunc == nil, "DiffSupppressFunc already defined, refusing to override")
+	contract.Assertf(field.Default != nil, "Default must be defined")
+	switch field.Type {
+	case schema.TypeBool:
+		defaultValue, ok := field.Default.(bool)
+		contract.Assertf(ok, "Default is not a boolean")
+		field.DiffSuppressFunc = func(k, oldValue, newValue string, d *schema.ResourceData) bool {
+			// oldValue must be undefined which is encoded as "" for bool
+			if oldValue != "" {
+				return false
+			}
+			// if newValue equals default value, suppress the diff
+			switch {
+			case defaultValue && newValue == "true":
+				return true
+			case !defaultValue && newValue == "false":
+				return true
+			default:
+				return false
+			}
+		}
+	case schema.TypeString:
+		defaultValue, ok := field.Default.(string)
+		contract.Assertf(ok, "Default is not a string")
+		field.DiffSuppressFunc = func(k, oldValue, newValue string, d *schema.ResourceData) bool {
+			// Here oldValue must be undefined.
```

Comment:
Noting the difficulty here, the underlying APIs are interesting.

Reply:
However, I think your use case can also be served by a TransformOutputs/TransformFromState hook, which should be able to distinguish between the two as it acts on the pulumi level.

Reply:
I don't think so. TransformFromState does not apply to import-y reads, and as to TransformOutputs I suspect if used to inject default values will also print them into the import program which this neatly avoids, I can check this out quick.

Reply:
>I don't think so. TransformFromState does not apply to import-y reads

Unfortunate - the docs suggest it does but it seems to only apply to the passed in state, not the one read from the provider.

Reply:
Checking up on this, looks like it does not in fact end up in defaults values being injected. Could be cleaner. I'll try harder.